code/game/objects/items/devices/holomap.dm

```diff
@@ -0,0 +1,70 @@
+// The station holomap, use it and you get a map of how the pipes should be.
+
+/obj/item/device/holomap
+	name = "holomap"
+	desc = "Displays the layout of the station's pipe and cable networks."
+
+	icon = 'icons/obj/device.dmi'
+	icon_state = "holomap"
+
+	var/list/image/showing = list()
+	var/client/viewing // Client that is using the device right now, also determines whether it's on or off.
+	//var/datum/delay_controller/delayer
+
+/obj/item/device/holomap/New()
+	..()
+	//delayer = new(0, ARBITRARILY_LARGE_NUMBER)
+
+/obj/item/device/holomap/Destroy()
+	//qdel(delayer)
+	//delayer = null
+
+	if (viewing)
+		viewing.mob.on_logout.Remove("\ref[src]:mob_logout")
+
+/obj/item/device/holomap/attack_self(var/mob/user)
+	if (viewing)
+		viewing.images -= showing
+		showing.Cut()
+		to_chat(user, "You turn off \the [src].")
+		viewing.mob.on_logout.Remove("\ref[src]:mob_logout")
+		viewing = null
+		return
+
+	if (!user.client) // delayer.blocked()
+		return
+
+	viewing = user.client
+	showing = get_images(get_turf(user), viewing.view)
+	viewing.images |= showing
+	//delayer.addDelay(2 SECONDS) // Should be enough to prevent lag due to spam.
+	user.on_logout.Add(src, "mob_logout")
+
+/obj/item/device/holomap/proc/mob_logout(var/list/args, var/mob/M)
+	if (viewing)
+		viewing.images -= showing
+		viewing = null
+
+	M.on_logout.Remove("\ref[src]:mob_logout")
+
+	visible_message("\The [src] turns off.")
+	showing.Cut()
+
+/obj/item/device/holomap/proc/get_images(var/turf/T, var/view)
+	. = list()
+	for (var/turf/TT in trange(view, T))
```

Comment:
shaving one proccall (granted a cheaper one because it's internal to DM) off of every iteration of the loop is something valuable, I'm just sadden that you chose to be dicks as opposed to simply saying "we don't wish to partake in these changes, don't share them again"